staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go

```diff
@@ -311,44 +321,214 @@ func updatePatchIfMissing(original, modified, patch map[string]interface{}, diff
 	}
 }
 
-// Returns a (recursive) strategic merge patch and a parallel deletion list if necessary.
+func validateMergeKeyInLists(mergeKey string, lists ...[]interface{}) error {
+	for _, list := range lists {
+		for _, item := range list {
+			m, ok := item.(map[string]interface{})
+			if !ok {
+				t := reflect.TypeOf(item)
+				return mergepatch.ErrBadArgType("map[string]interface{}", t.Kind().String())
+			}
+			if _, ok := m[mergeKey]; !ok {
+				return mergepatch.ErrNoMergeKey(m, mergeKey)
+			}
+		}
+	}
+	return nil
+}
+
+// sortSliceByMultipleOrders sort `patch` list by patchOrder and
+// sort `serverOnly` list by serverOrder.
+// Then it merges the 2 sorted lists by serverOrder with best effort.
+// It guarantee the relative order in the patch list and in the serverOnly list is kept.
+func sortSliceByMultipleOrders(patch, serverOnly, patchOrder, serverOrder []interface{}, mergeKey string, kind reflect.Kind) ([]interface{}, error) {
+	patch, err := sortSliceByOneOrder(patch, patchOrder, mergeKey, kind)
+	if err != nil {
+		return nil, err
+	}
+	serverOnly, err = sortSliceByOneOrder(serverOnly, serverOrder, mergeKey, kind)
+	if err != nil {
+		return nil, err
+	}
+	all := mergeSortedSlice(serverOnly, patch, serverOrder, mergeKey, kind)
+
+	return all, nil
+}
+
+// mergeSortedSlice merges the 2 sorted lists by serverOrder with best effort.
```

Comment:
> The relative order of two items are determined by the following order:
>- relative order in the $setElementOrder if both items are present
>- else relative order in the patch if both items are present
>- else relative order in the server-side list if both items are present
>- else append to the end

We are tying to sort the list according to this [proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/preserve-order-in-strategic-merge-patch.md#proposed-change).
But there may be ordering conflict. e.g.
live list: `["a", "x", "b", "y", "c", "z"]`
patch list: `["c", "b", "a"]`
There is no way to make a list satisfy:
- `a` comes before `x`
- `x` comes before `b`
- `b` comes before `a`

In this PR's implementation, we can guarantee the relative order in the patch list and the relative order between server-only items.
We have 2 sorted list: sorted patch list and sorted server-only items list.
We are trying to merge these 2 lists.
Please check [sortSliceByMultipleOrders](https://github.com/kubernetes/kubernetes/pull/45980/files#diff-101008d96c4444a5813f7cb6b54aaff6R344).